add search to featureFlag

samiramkr opened the pull request:
Signed-off-by: Samira Mokaram <samiram@spotify.com>

## Hey, I just made a Pull Request!

This PR will add search to FeatureFlags. When there is a long list of feature flags it's hard to find them without a search.


![Screenshot 2021-09-17 at 10 19 48](https://user-images.githubusercontent.com/15106494/133749625-55036ab0-53ec-4ed5-be7b-03d01f4ceaff.png)

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

freben commented on plugins/user-settings/src/components/FeatureFlags/UserSettingsFeatureFlags.tsx:
```diff
@@ -35,6 +35,7 @@ export const UserSettingsFeatureFlags = () => {
   );
 
   const [state, setState] = useState<Record<string, boolean>>(initialFlagState);
+  const [searchInput, setSearchInput] = useState<string>('');
```
Can we use the word "filter" throughout this PR instead of "search"? I think that's consistent with expectations and also doesn't collide with the larger concept of search in Backstage. The PR title really confused me for a while because of this :)

freben commented on plugins/user-settings/src/components/FeatureFlags/UserSettingsFeatureFlags.tsx:
```diff
@@ -62,16 +63,25 @@ export const UserSettingsFeatureFlags = () => {
-              enabled={enabled}
-              toggleHandler={toggleFlag}
-            />
+            featureFlag.name.includes(searchInput) && (
```
future improvement / over-engineering idea: `searchInput.split(/\s/).map(part => part.trim().toLowerCase()).filter(Boolean)` (outside of this `.map` loop) and then check that `featureFlag.name.toLowerCase()` contains ALL of those parts :)

Fox32 commented:
I wonder if it's possible to place the search box in the top right of the card header? Right now it looks a bit displaced, and also not that obvious that it is a search/filter control at all.

samiramkr commented:
> I wonder if it's possible to place the search box in the top right of the card header? Right now it looks a bit displaced, and also not that obvious that it is a search/filter control at all.

To be honest, UX-wise, I have no idea. Maybe @katz95 can help us?

timbonicus commented on plugins/user-settings/src/components/FeatureFlags/UserSettingsFeatureFlags.tsx:
```diff
@@ -62,16 +63,25 @@ export const UserSettingsFeatureFlags = () => {
       <List dense>
-        {featureFlags.map(featureFlag => {
+        <ListItem>
+          <TextField
```
Can we set a threshold for this to appear (e.g. `featureFlags.length >= 10`)? I’m thinking it will be common to have only a few active feature flags.

samiramkr replied:
I'm not sure if I understood correctly! On our internal app, we have many `featureFlags`. It will be confusing for users if they can't see all their active feature flags!

timbonicus replied:
Oh definitely, all feature flags should be visible. I meant that the search box itself should only appear if enough feature flags exist to make search needed. If there's only 3 flags, for example, a search may add more clutter than actual utility. 😄

samiramkr replied:
aha, that makes sense 👍

samiramkr replied:
I added to show it if items are more than x, but it's working because when the number of items goes under the x, the filter is gone!

samiramkr commented:
I moved the Filter into the card's header.
![Screenshot 2021-09-17 at 18 28 18](https://user-images.githubusercontent.com/15106494/133822523-9bc54bb5-e5ce-4191-aa4a-e0648e158b4e.png)

I added a clear button, which is rendered when a value is in input.
![Screenshot 2021-09-17 at 18 28 27](https://user-images.githubusercontent.com/15106494/133822634-98923068-82cc-4844-b100-d5bdc14adf44.png)

![Screenshot 2021-09-17 at 18 28 37](https://user-images.githubusercontent.com/15106494/133822649-dccc79ba-eedd-4dbb-9362-2ac69a20d289.png)

timbonicus commented:
Hmm, it definitely looks odd in the "subtitle" area of the header. I think @Fox32 was suggesting a right-aligned search similar to the [Table search shown in Storybook](https://backstage.io/storybook/?path=/story/data-display-table--default-table).

Fox32 commented:
> Hmm, it definitely looks odd in the "subtitle" area of the header. I think @Fox32 was suggesting a right-aligned search similar to the [Table search shown in Storybook](https://backstage.io/storybook/?path=/story/data-display-table--default-table).

Yeah, exactly what I meant! A screenshot would have said more than 1000 words 🤦‍♂️😅

samiramkr commented:
> Hmm, it definitely looks odd in the "subtitle" area of the header. I think @Fox32 was suggesting a right-aligned search similar to the [Table search shown in Storybook](https://backstage.io/storybook/?path=/story/data-display-table--default-table).

I got it 👍

samiramkr commented:
Moved to the right
![Screenshot 2021-09-21 at 11 43 54](https://user-images.githubusercontent.com/15106494/134149249-71983d99-2372-48d6-b087-706894fef3f9.png)

samiramkr commented:
The filter is not shown for less than 10 items

![Screenshot 2021-09-21 at 13 21 14](https://user-images.githubusercontent.com/15106494/134161881-7236d9cc-2130-4a36-9e49-c09ba8cabf65.png)

freben commented on plugins/user-settings/src/components/FeatureFlags/UserSettingsFeatureFlags.tsx:
```diff
@@ -59,15 +68,67 @@ export const UserSettingsFeatureFlags = () => {
+            InputProps={{
+              ...(filterInput.length && {
+                endAdornment: (
+                  <React.Fragment>
```
don't need this extra fragment

freben commented on plugins/user-settings/src/components/FeatureFlags/UserSettingsFeatureFlags.tsx:
```diff
@@ -59,15 +68,67 @@ export const UserSettingsFeatureFlags = () => {
+                endAdornment: (
+                  <React.Fragment>
+                    <IconButton
+                      aria-label="clear filterInput's value"
```
```suggestion
                      aria-label="Clear filter"
```

freben commented on plugins/user-settings/src/components/FeatureFlags/UserSettingsFeatureFlags.tsx:
```diff
@@ -59,15 +68,67 @@ export const UserSettingsFeatureFlags = () => {
           return (
             <FlagItem
-              key={featureFlag.name}
+              key={index + 1}
```
I think this was actually right before this - the name was probably a perfect key to use, since the key is used to detect items moving around and saving re-renders. Did you change this for some particular reason?

freben reviewed (commented):
Almost done! Looks good
